[SPARK-9792] Make DenseMatrix equality semantical

gglanzani opened the pull request:
Before, you could have this code

```
A = SparseMatrix(2, 2, [0, 2, 3], [0], [2])
B = DenseMatrix(2, 2, [2, 0, 0, 0])

B == A  # False
A == B  # True
```

The second would be `True` as `SparseMatrix` already checks for semantic
equality. This commit changes `DenseMatrix` so that equality is
semantical as well.

## What changes were proposed in this pull request?

Better semantic equality for DenseMatrix

## How was this patch tested?

Unit tests were added, plus manual testing. Note that the code falls back to the old behavior when `other` is not a SparseMatrix.

AmplabJenkins commented:
Can one of the admins verify this patch?

WeichenXu123 commented on python/pyspark/mllib/linalg/__init__.py:
```diff
@@ -1131,14 +1131,21 @@ def __getitem__(self, indices):
+            return np.all(self_values == other_values)
+
+        if isinstance(other, SparseMatrix):
+            return _eq(self, other)
```
This implementation is incorrect here I think.
You flatten 2D-arrays for both side, and then compare the two flattened array. it will result in `M*N` matrix equals `N*M` matrix when their values are the same.

So you need move the judgement of
```
if  (self.numRows != other.numRows or		                  
      self.numCols != other.numCols):
     return False
```
into front.
Thanks!

WeichenXu123 commented:
ping @gglanzani This bug need fixed ASAP. Can you update code when you're free ? Thanks.

WeichenXu123 commented:
@gglanzani And you the `ml.linalg.DenseMatrix` looks have the same bug. Can you also update it ?

gglanzani commented:
@WeichenXu123 Done. Let me know.

WeichenXu123 commented on python/pyspark/ml/linalg/__init__.py:
```diff
@@ -976,14 +976,20 @@ def __getitem__(self, indices):
             return self.values[i + j * self.numRows]
 
     def __eq__(self, other):
+        def _eq(self, other):
```
Now `_eq` won't be called in multiple places, we can remove the `_eq` definition, make code simpler.

WeichenXu123 commented on python/pyspark/mllib/linalg/__init__.py:
```diff
@@ -1131,14 +1131,20 @@ def __getitem__(self, indices):
             return self.values[i + j * self.numRows]
 
     def __eq__(self, other):
+        def _eq(self, other):
```
Ditto.

gglanzani commented:
@WeichenXu123 Done again!

WeichenXu123 commented on python/pyspark/mllib/linalg/__init__.py:
```diff
@@ -1131,14 +1131,17 @@ def __getitem__(self, indices):
     def __eq__(self, other):
+        if isinstance(other, SparseMatrix):
+            return np.all(self.toArray() == other.toArray())
         if (not isinstance(other, DenseMatrix) or
```
ditto.

WeichenXu123 commented on python/pyspark/ml/linalg/__init__.py:
```diff
@@ -976,14 +976,18 @@ def __getitem__(self, indices):
+
+        if isinstance(other, SparseMatrix):
+            return np.all(self.toArray() == other.toArray())
         if (not isinstance(other, DenseMatrix) or
```
The condition not `isinstance(other, DenseMatrix)` is useless. remove it.

WeichenXu123 commented:
@gatorsmile Add this to white list!

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

gatorsmile commented:
cc @WeichenXu123 @yanboliang

SparkQA commented:
**[Test build #85738 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85738/testReport)** for PR 17968 at commit [`311c94a`](https://github.com/apache/spark/commit/311c94a3d608b0b86f3ce39415639ec260e5af37).

SparkQA commented:
**[Test build #85738 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85738/testReport)** for PR 17968 at commit [`311c94a`](https://github.com/apache/spark/commit/311c94a3d608b0b86f3ce39415639ec260e5af37).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85738/
Test PASSed.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins OK to test
Are you still actively wortking on this and if so would you update it to master?

holdenk commented:
jenkins ok to test

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7988/
Test PASSed.

SparkQA commented:
**[Test build #102408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102408/testReport)** for PR 17968 at commit [`311c94a`](https://github.com/apache/spark/commit/311c94a3d608b0b86f3ce39415639ec260e5af37).

SparkQA commented:
**[Test build #102408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102408/testReport)** for PR 17968 at commit [`311c94a`](https://github.com/apache/spark/commit/311c94a3d608b0b86f3ce39415639ec260e5af37).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102408/
Test FAILed.

gglanzani committed:
[SPARK-9792] Make DenseMatrix equality semantical
Before, you could have this code

```
A = SparseMatrix(2, 2, [0, 2, 3], [0], [2])
B = DenseMatrix(2, 2, [2, 0, 0, 0])

B == A  # False
A == B  # True
```

The second would be `True` as `SparseMatrix` already checks for semantic
equality. This commit changes `DenseMatrix` so that equality is
semantical as well.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8029/
Test PASSed.

SparkQA commented:
**[Test build #102450 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102450/testReport)** for PR 17968 at commit [`1f8b8ff`](https://github.com/apache/spark/commit/1f8b8ff2a548cc23ef2ca82db0cdd5c1c6258f86).

SparkQA commented:
**[Test build #102450 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102450/testReport)** for PR 17968 at commit [`1f8b8ff`](https://github.com/apache/spark/commit/1f8b8ff2a548cc23ef2ca82db0cdd5c1c6258f86).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102450/
Test FAILed.

gglanzani commented:
@holdenk Yep, rebased. Sorry for the failed Python style test, fixed now.

SparkQA commented:
**[Test build #102453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102453/testReport)** for PR 17968 at commit [`eab84ef`](https://github.com/apache/spark/commit/eab84ef074f309227a226b130bb093f9896fd3ab).

SparkQA commented:
**[Test build #102453 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102453/testReport)** for PR 17968 at commit [`eab84ef`](https://github.com/apache/spark/commit/eab84ef074f309227a226b130bb093f9896fd3ab).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102453/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8032/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8033/
Test PASSed.

SparkQA commented:
**[Test build #102454 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102454/testReport)** for PR 17968 at commit [`c79250f`](https://github.com/apache/spark/commit/c79250fbfbfc169de98751317cb075b8f1d966cf).

SparkQA commented:
**[Test build #102454 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102454/testReport)** for PR 17968 at commit [`c79250f`](https://github.com/apache/spark/commit/c79250fbfbfc169de98751317cb075b8f1d966cf).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102454/
Test FAILed.

holdenk commented:
Looks like unrelated test failure.
Jenkins retest this please.

holdenk commented on python/pyspark/ml/linalg/__init__.py:
```diff
@@ -980,14 +980,14 @@ def __getitem__(self, indices):
-                self.numCols != other.numCols):
+        if isinstance(other, SparseMatrix):
+            return np.all(self.toArray() == other.toArray())
+        if (self.numRows != other.numRows or self.numCols != other.numCols):
```
I think it would make sense to move this check above L983? We canm check the number of rows and columns faster than converting a `SparseMatrix` to an array.

holdenk commented on python/pyspark/mllib/linalg/__init__.py:
```diff
@@ -1135,14 +1135,14 @@ def __getitem__(self, indices):
-                self.numCols != other.numCols):
+        if isinstance(other, SparseMatrix):
+            return np.all(self.toArray() == other.toArray())
+        if (self.numRows != other.numRows or self.numCols != other.numCols):
```
Same comment as in the Ml version

holdenk commented:
Jenkins retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8691/
Test PASSed.

SparkQA commented:
**[Test build #103246 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103246/testReport)** for PR 17968 at commit [`c79250f`](https://github.com/apache/spark/commit/c79250fbfbfc169de98751317cb075b8f1d966cf).

SparkQA commented:
**[Test build #103246 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103246/testReport)** for PR 17968 at commit [`c79250f`](https://github.com/apache/spark/commit/c79250fbfbfc169de98751317cb075b8f1d966cf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103246/
Test PASSed.

holdenk commented:
Just a gentle ping @gglanzani  if you have time, sorry for the slow review process.

SparkQA commented:
**[Test build #103365 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103365/testReport)** for PR 17968 at commit [`69db472`](https://github.com/apache/spark/commit/69db472326026c13b62ba9c5f30952a04cbf5f52).

gglanzani commented:
@holdenk done (Github tries to be smart so it moved the comments, but the logic indeed changed as you suggested).

SparkQA commented:
**[Test build #103365 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103365/testReport)** for PR 17968 at commit [`69db472`](https://github.com/apache/spark/commit/69db472326026c13b62ba9c5f30952a04cbf5f52).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103365/
Test PASSed.

holdenk reviewed (approved):
LGTM thanks for fixing this.

holdenk commented:
LGTM, @gglanzani if you can sign up for a JIRA account on https://issues.apache.org/jira/browse/SPARK-9792 I can assign this JIRA to you.

holdenk commented:
Merged to master

gglanzani commented:
@holdenk thanks. I’m `lanzani` on Jira
